src/app/ryderserial/ryder-serial/output_parser.ts

```diff
@@ -0,0 +1,126 @@
+import { RESPONSE_OUTPUT, RESPONSE_OUTPUT_END, RESPONSE_ESC_SEQUENCE } from "./responses";
+
+/* The kind of output received from the Ryder device. */
+enum OutputKind {
+  /* A single-byte response. */
+  Single,
+  /* A multi-byte response. */
+  Multi,
+}
+
+/*
+ * An online parser for output from the Ryder device. Handles both single- and multi-byte output,
+ * works with partial data supplied as it is received, and can indicate when the output is
+ * completed.
+ */
+export class OutputParser {
+  /* A buffer to store the parsed output. */
+  private buffer: string;
+  private kind?: OutputKind;
+  /* Whether the output is completed (i.e., no more data is expected). */
+  private completed: boolean;
+  /* Whether to escape the next byte received. */
+  private escape_next_byte: boolean;
+
+  /* Initializes an empty output parser, ready to receive data. */
+  constructor() {
+    this.buffer = '';
+    this.completed = false;
+    this.escape_next_byte = false;
+  }
+
+  /*
+   * Adds data to the output parser and returns the output if it is complete.
+   *
+   * @param data The data to add to the output, interpreted as a byte array.
+   */
+  add_data(data: string): string | undefined {
+    // TODO: unit tests
+    if (this.completed) {
+      throw new Error('`add_data` called on completed output');
+    }
+
+    // If adding no data, do nothing
+    if (data.length === 0) {
+      return;
+    }
+
+    // If this is the first data received, check whether it is single- or multi-byte output
+    let is_first_byte = this.kind === undefined;
+    if (is_first_byte) {
+      if (data.charCodeAt(0) === RESPONSE_OUTPUT) {
+        // This is the beginning of a multi-byte output
+        this.kind = OutputKind.Multi;
+      } else {
+        // All other bytes come by themselves
+        this.kind = OutputKind.Single;
+      }
+    }
+
+    // Whether there was more data than expected for a single command's output
+    let extra_data = false;
+
+    if (this.kind === OutputKind.Single) {
+      this.buffer += data[0];
+      this.completed = true;
+
+      // Detect if there is more data than expected
+      if (data.length > 1) {
+        extra_data = true;
+      }
+    } else if (this.kind === OutputKind.Multi) {
+      for (let i = 0; i < data.length; i++) {
+        // Detect if there is more data than expected
+        if (this.completed) {
+          extra_data = true;
+          break;
+        }
+
+        const byte = data.charCodeAt(i);
+
+        if (this.escape_next_byte) {
+          // Escaped bytes are added as-is
+          this.buffer += String.fromCharCode(byte);
```

Comment:
I think the rewrite is a good opportunity to move away from `string`. It would be much better if `this.buffer` is a `Uint8Array`.

Reply:
Agreed, I'll replace it. Though `ryder-utils` and perhaps the wallet itself seem to require strings, so I'll just add a conversion later on.

Reply:
Yeah let's update RyderSerial to only use `Uint8Array` and if some legacy code uses strings we will convert the buffer to a string there and update gradually. I think a refactor shouldn't take that much time.